### PR TITLE
Implement reference for lerp

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87182
* #87183
* #87205
* #87204
* #87203
* __->__ #87424

We follow the vectorised CPU implementation for numerical accuracy

cc @ezyang @mruberry @ngimel @Lezcano @fdrocha

### CODE REVIEW

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -4025,6 +4026,29 @@ def arange(
...
 
 
+@register_decomposition(torch.ops.aten.lerp)
+@out_wrapper()
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("start", "end", "weight"),
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+)
+def lerp(start: Tensor, end: Tensor, weight: Union[Tensor, NumberType]):
+    check(start.dtype == end.dtype, lambda: f"expected dtype {start.dtype} for `end` but got dtype {end.dtype}")
```
Comment (ngimel) [inline torch/_refs/__init__.py:4036]: too late for that after type promotion wrapper?

Comment (lezcano) [reply torch/_refs/__init__.py:4036]: I see this as "the reference implements an extended behaviour over the PyTorch op", as we support a few more combinations of dtypes.

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -4060,6 +4061,36 @@ def arange(
...
+    if isinstance(weight, Number):
+        weight = start.new_full((), weight)  # type: ignore[arg-type]
+    else:
+        check(
+            start.dtype == weight.dtype,
+            lambda: f"expected dtype {start.dtype} for `weight` but got dtype {weight.dtype}",  # type: ignore[union-attr]
+        )
+    assert isinstance(weight, Tensor)  # mypy
+    # We implement it this way for numerical stability. We assume (in the stability optimisation)
+    # that 0 <= weight <= 1. We take the abs to deal with comples numbers
```
Comment (ezyang) [inline torch/_refs/__init__.py:4084]: Complex

**Review by ezyang — APPROVED** [review]

**Comment by lezcano** [comment]
@pytorchbot merge

I'll fix the errata in the next PR of the stack

File: torch/_refs/__init__.py (hunk 1 above)
Comment (ngimel) [reply torch/_refs/__init__.py]: Ok, but this check in reference is then pointless, the inputs it's getting are all the same type.

Comment (lezcano) [reply torch/_refs/__init__.py]: ah, I see what you mean. I'll clean this up later in this stack.

Comment (lezcano) [reply torch/_refs/__init__.py]: Added a clean-up in https://github.com/pytorch/pytorch/pull/87183

**Pull request closed without merging**